Add agent config option to control Window Counter refresh rate

steveny91 opened the pull request:
### What does this PR do?
Add a param to the agent to control how often window counters should be refreshed.

Linked PR:
https://github.com/DataDog/datadog-agent/pull/13273

clarkb7 commented:
Since there is only one `WindowsPerformanceObjectRefresher` instance that is shared among all checks that inherit from `PerfCountersBaseCheck` there is only ever one interval to be set.

Is it possible to put this code in `WindowsPerformanceObjectRefresher` instead of `PerfCountersBaseCheck` ?
```python
self.interval = datadog_agent.get_config('windows_counter_refresh_interval')
```

If not, maybe suppress the "Windows Counters refresh interval set to" log message if the interval isn't changing, so we don't get the same message from every check.

iglendd commented on datadog_checks_base/datadog_checks/base/checks/windows/perf_counters/base.py:
```diff
@@ -40,7 +49,10 @@ def __init__(self, name, init_config, instances):
         self.check_initializations.append(self.create_connection)
-        self.check_initializations.append(self.setup_refresher)
+
+        if self.interval > 0:
```
That is good, we can suppress refresshing

iglendd commented on datadog_checks_base/datadog_checks/base/checks/windows/perf_counters/refresh.py:
```diff
@@ -67,3 +66,7 @@ def remove_server(self, server):
     def log_server_count(self, server):
         self.logger.info('Refresh counter set to %d for server: %s', self.servers[server], server)
+
+    def add_interval(self, interval):
```
should it be called set_interval instead?

steveny91 replied:
I removed this function as I was addressing a comment from @clarkb7

iglendd commented on datadog_checks_base/datadog_checks/base/checks/windows/perf_counters/base.py:
```diff
@@ -25,6 +30,10 @@ class PerfCountersBaseCheck(AgentCheck):
         super().__init__(name, init_config, instances)
 
+        self.interval = 60
+        if datadog_agent and datadog_agent.get_config('windows_counter_refresh_interval'):
```
Do we need to document it in some wayh

steveny91 replied:
Yes, we will add the documentation for this feature and the nuances once this feature is in.

steveny91 commented:
@clarkb7 what you said made a lot of sense. Thanks for the feedback. I implemented your suggestion. Feel free to give it a look.

I also added a linked PR for the parameter that will be added in the datadog.yaml

clarkb7 reviewed (approved):
LGTM

jose-manuel-almaza reviewed (commented):
I imagine this value of time.sleep(2) causes a conflict if I set 'windows_counter_refresh_interval' to 1 (I do not know if this value makes sense, but I can)
It won't work as I expect, will it?

ofek commented on datadog_checks_base/datadog_checks/base/checks/windows/perf_counters/refresh.py:
```diff
@@ -9,17 +9,24 @@
         super(WindowsPerformanceObjectRefresher, self).__init__(name=name)
 
+        self.interval = 60
+        if datadog_agent and datadog_agent.get_config('windows_counter_refresh_interval'):
```
Checking for truthyness prevents the ability to set to zero

iglendd replied:
Theoretically a good find but would not this prevent refreshes thread in a first place https://github.com/DataDog/integrations-core/pull/12665/files#diff-77f24038c0b7c7947277e93e51cf208efa6f8ddfbed47213c566533a6b0ba4f7R46-R47?

steveny91 replied:
Thanks for that. I edited it just to keep things more concise.

ofek commented on datadog_checks_base/datadog_checks/base/checks/windows/perf_counters/refresh.py:
```diff
@@ -9,17 +9,24 @@
         super(WindowsPerformanceObjectRefresher, self).__init__(name=name)
 
+        self.interval = 60
+        if datadog_agent and not datadog_agent.get_config('windows_counter_refresh_interval') == None:
```
Pythonic nit

```suggestion
        if datadog_agent and datadog_agent.get_config('windows_counter_refresh_interval') is not None:
```

steveny91 replied:
Added and 📔

sarah-witt reviewed (approved):
🎉

FlorentClarret commented:
Unfrozen
